Properly read SQL-compatible segments in default-value mode.

gianm opened the pull request:
Main changes:

1) Dictionary-encoded and front-coded string columns: in default-value
   mode, detect cases where a dictionary has the empty string in it, then
   either combine it with null (if null is present) or replace it with
   null (if null is not present).

2) Numeric nullable columns: in default-value mode, ignore the null
   value bitmap. This causes all null numbers to be read as zeroes.

Testing strategy:

1) Add a mmappedWithSqlCompatibleNulls case to BaseFilterTest that
   writes segments under SQL-compatible mode, and reads them under
   default-value mode.

2) Unit tests for the new wrapper classes (CombineFirstTwoEntriesIndexed,
   CombineFirstTwoValuesColumnarInts, CombineFirstTwoValuesColumnarMultiInts,
   CombineFirstTwoValuesIndexedInts).

gianm commented:
Currently a draft PR since I haven't written unit tests for the new helper classes yet. I'm interested in feedback on the approach though.

gianm commented:
There is some performance overhead here for the case where a server in default-value mode is reading a segment that was written in SQL-compatible mode. However, I don't really see a way around this. The dictionaries do need to be adjusted and it will add some extra overhead.

There's no overhead for default-value mode reading default-value-mode-written segments, or SQL-compatible mode reading any kind of segments.

gianm commented:
Added tests and marked this PR as ready for review.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/column/StringFrontCodedDictionaryEncodedColumn.java:
```diff
@@ -68,12 +68,12 @@ public class StringFrontCodedDictionaryEncodedColumn implements DictionaryEncode
   @Nullable
   private final ColumnarMultiInts multiValueColumn;
-  private final FrontCodedIndexed utf8Dictionary;
+  private final Indexed<ByteBuffer> utf8Dictionary;
```
i think this and `ScalarStringDictionaryEncodedColumn` can now be combined but I can do that in a follow-up PR

gianm replied:
I did this in the latest patch.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/serde/ReplaceFirstValueWithNullIndexed.java:
```diff
@@ -0,0 +1,131 @@
+ *
+ * @see NullHandling#mustReplaceFirstValueWithNullInDictionary(Indexed)
+ */
+public class ReplaceFirstValueWithNullIndexed<T> implements Indexed<T>
```
is this primarily so that callers don't have to worry about emptyToNullIfNeeded/nullToEmptyIfNeeded for the case where the dictionary had "" but no nulls?

clintropolis reviewed (approved):
I think the 'auto' string column should probably (unfortunately) use this stuff too, but I think it can be done as a follow-up, since I also think that it can be totally combined with the front-coded string column (which is no longer specific to front-coding after this PR, rather its a string column which only has a utf8 buffer dictionary).

gianm commented:
Pushed up a patch that addresses these comments:

    1) Read bitmaps even if we don't retain them.
    2) Combine StringFrontCodedDictionaryEncodedColumn and ScalarStringDictionaryEncodedColumn.
